src/morphology.rs

```diff
@@ -349,6 +351,563 @@ pub fn close_mut(image: &mut GrayImage, norm: Norm, k: u8) {
+                .collect(),
+        }
+    }
+
+    fn apply<'a, 'b: 'a, 'c: 'a>(
+        &'c self,
+        image: &'b GrayImage,
+        x: u32,
+        y: u32,
+    ) -> impl Iterator<Item = &'a Luma<u8>> {
+        self.elements
+            .iter()
+            .map(move |(i, j)| (x as i64 + *i as i64, y as i64 + *j as i64))
+            .filter(move |(i, j)| {
+                0 <= *i && *i < image.width() as i64 && 0 <= *j && *j < image.height() as i64
+            })
+            .map(move |(i, j)| image.get_pixel(i as u32, j as u32))
+    }
+}
+
+/// computes the morphologic dilation of the input image with the given mask
+///
+/// for each input pixel, the output pixel will be the maximum of all pixels included
+/// in the mask at that position.
+///
+/// # Examples
+/// ```
+/// # extern crate image;
+/// # #[macro_use]
+/// # extern crate imageproc;
+/// # fn main() {
+/// use image::GrayImage;
+/// use imageproc::morphology::{Mask, grayscale_dilate};
+///
+/// let image = gray_image!(
+///     7,   0,   0,   0,   0,   0;
+///     0,   0,   0,   0,   0,   0;
+///     0,   0,  99,   0,   0,   0;
+///     0,   0,   0,   0,   0,   0;
+///     0,   0,   0,   0,   0, 222
+/// );
+///
+/// // using a diamond mask
+/// let diamond_dilated = gray_image!(
+///     7,   7,   0,   0,   0,   0;
+///     7,   0,  99,   0,   0,   0;
+///     0,  99,  99,  99,   0,   0;
+///     0,   0,  99,   0,   0, 222;
+///     0,   0,   0,   0, 222, 222
+/// );
+///
+/// assert_pixels_eq!(grayscale_dilate(&image, &Mask::diamond(1)), diamond_dilated);
+///
+/// // using a disk mask
+/// let disk_dilated = gray_image!(
+///    99,  99,  99,  99,  99,   0;
+///    99,  99,  99,  99,  99, 222;
+///    99,  99,  99, 222, 222, 222;
+///    99,  99,  99, 222, 222, 222;
+///    99,  99, 222, 222, 222, 222
+/// );
+///
+/// assert_pixels_eq!(grayscale_dilate(&image, &Mask::disk(3)), disk_dilated);
+///
+/// // using a square mask
+/// let square_dilated = gray_image!(
+///     7,   7,   0,   0,   0,   0;
+///     7,  99,  99,  99,   0,   0;
+///     0,  99,  99,  99,   0,   0;
+///     0,  99,  99,  99, 222, 222;
+///     0,   0,   0,   0, 222, 222
+/// );
+///
+/// assert_pixels_eq!(grayscale_dilate(&image, &Mask::square(1)), square_dilated);
+///
+///
+/// // using an arbitrary mask
+/// // the center of the mask is the '4' here, cause it's coordinates are (0,1)
+/// let column_mask = Mask::from_image(
+///     &gray_image!(
+///         8;
+///         1;
+///         9;
+///         2
+///     ),  
+///     0,  1
+/// );
+///
+/// let column_dilated = gray_image!(
+///     7,   0,  99,   0,   0,   0;
+///     7,   0,  99,   0,   0,   0;
+///     0,   0,  99,   0,   0, 222;
+///     0,   0,  99,   0,   0, 222;
+///     0,   0,   0,   0,   0, 222
+/// );
+///
+/// assert_pixels_eq!(grayscale_dilate(&image, &column_mask), column_dilated);
+/// # }
+/// ```
+pub fn grayscale_dilate(image: &GrayImage, mask: &Mask) -> GrayImage {
+    #[cfg(feature = "rayon")]
```

Comment:
We can do it later actually

Reply:
i am very interested in any different ideas on this.

i have been thinking about optimization, and i think that ultimately it would be idea to parallelize on image_line/mask_line pairs for cache locality, but that will be tested and changed later.

Reply:
I just want users to know if the function will consume the resources of the rayon thread pool or not. Our template matching for example, we have `match_template` and `match_template_parallel`, not 1 function for both depending on the feature.

Reply:
doing it in another pr sounds great, if we can have finalize a working foundation and work on it afterward that would be great

Reply:
Imagine I have a lot of clients (webserver) or `batch` of work. `inputs.par_iter().map(match_template)` gonna be faster than `inputs.par_iter().map(match_template_parallel)`.

Reply:
So I think you can just delete it.

Reply:
just did that in my last commit